types/util.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+// http://cavaliercoder.com/blog/optimized-abs-for-int64-in-go.html
+func abs(n int64) int64 {
+	y := n >> 63
+	return (n ^ y) - y
+}
+
+// uintSizeTable is used as a table to do comparison to get uint length is faster than doing loop on division with 10
+var uintSizeTable = [21]uint64{
+	0, // redundant 0 here, so to make function lenOfUint64Fast to count from 1 and return i directly
+	9, 99, 999, 9999, 99999,
+	999999, 9999999, 99999999, 999999999, 9999999999,
+	99999999999, 999999999999, 9999999999999, 99999999999999, 999999999999999,
+	9999999999999999, 99999999999999999, 999999999999999999, 9999999999999999999,
+	maxUint,
+} // maxUint is 18446744073709551615 and it has 20 digits
+
+// LenOfUint64Fast efficiently calculate the string character lengths of an uint64 as input
+func LenOfUint64Fast(x uint64) int {
+	i := 1
+	// Reduce comparison time for large numbers
+	// Optimized for input x mostly range between 0 to 99999
+	if x > 9999 {
```

Comment:
return i ?

Reply:
@ngaut not sure what you mean here? `i` cannot be returned here as we don't know what is the length of x yet, we only know it is longer than 4, so we set "i = 5" and search from there, instead from 0, to 1 , to 2 ... to 5 and then continue.

Reply:
My bad, sorry for misunderstood